### PR TITLE
Improve JSONL handling

### PR DESCRIPTION
**Pull request by daviewales**
Work in progress for #1726

### CODE REVIEW

**Review by saulpw — APPROVED** [review]

**Comment by daviewales** [comment]
I was thinking of bundling the following into this PR, but also happy if you want to merge this and I'll send another PR when the rest is ready:

- [x] Add warning when saving `jsonl` file with no rows
- [x] Add JSONL arrays loader
- [x] Add JSONL arrays saver
- [x] Add option to save JSONL as arrays
- [x] Add extended warning which hints to use JSONL arrays format to preserve headers
- [x] Add tests

**Comment by saulpw** [comment]
We can sit on it for a bit while you're actively developing, see if we want to tweak the message or anything.

**Comment by daviewales** [comment]
@saulpw, I think I'm ready for you to look at this now. I've made a few potentially controversial choices, so keen to hear your thoughts! A summary of my questionable choices follows:

- JSONL arrays get their own sheet type, inheriting from SequencSheet
- JSONL arrays get their own file extension (`.jsonla`)
- Loading as JSONL arrays automatically falls back to the default JSONL loader if the file doesn't include a JSONL arrays header
- I've split the `json_save_arrays` option in two: `json_save_arrays` and `json_load_arrays`. The first will always use the arrays format for `.jsonl` files, and the second will always attempt use the arrays format for `.jsonl` files, falling back to the default if the header is not detected.

From a user perspective, there is _no change_ if they save as `.jsonl` and have the new options unset.

I haven't added any tests yet, so I'll look into that.

**Comment by daviewales** [comment]
**EDIT**: Sorry, I think I understand now...

~~Perhaps I don't understand how testing works.
But it looks to me that the existing `save-json.vd` test doesn't actually save anything?
https://github.com/saulpw/visidata/blob/develop/tests/save-json.vd~~

**Comment by daviewales** [comment]
Hmm... Testing has revealed a bug when saving as `.jsonla`.

**Comment by daviewales** [comment]
I think we're good now.

**Edit**: Helps to commit the sample data...

**Edit 2**: Removed some vestigial code and force pushed

File: tests/load-jsonla.vdj
Comment (saulpw) [inline tests/load-jsonla.vdj]: Let's remove the column sorting; we should be checking that the columns come in, in the order of the array.

File: tests/save-jsonla.vdj
Comment (saulpw) [inline tests/save-jsonla.vdj:2]: Let's type the columns at least as `int` and `float`, so we can make sure they're being saved as such since .jsonla supports numbers.

File: visidata/loaders/json.py
Comment (saulpw) [inline visidata/loaders/json.py]: Use `vd.options` here; only use `<SheetClass>.options` for overriding (setting) defaults on SheetClass.

File: visidata/loaders/json.py
Comment (saulpw) [inline visidata/loaders/json.py]: Let's remove these extra `vd.status()` in this function

File: visidata/loaders/json.py
Comment (saulpw) [inline visidata/loaders/json.py]: We don't need these options, I think.  For saving, using the filetype `jsonla` should take care of it.  For loading, `open_jsonla` already checks the first line for an array type and falls back to `open_json` if it's not.  We'll make a better structure for guessing filetypes ultimately, but until then it should be workable without extra options.

**Review by saulpw — COMMENTED** [review]
Thanks for pulling this together, @daviewales !  I had a few comments, and if we address those I think we should be able to merge it.

File: tests/save-jsonla.vdj
Comment (daviewales) [reply tests/save-jsonla.vdj:2]: Hmm... I just tried to play this command log and this happened:
Screen Shot 2023-02-15 at 9 35 14 pm

From what I can tell, it glitches when asking whether to overwrite an existing file.
Otherwise the replay works without issue.

File: visidata/loaders/json.py
Comment (daviewales) [reply visidata/loaders/json.py]: Removed.

File: visidata/loaders/json.py
Comment (daviewales) [reply visidata/loaders/json.py]: Done.

File: visidata/loaders/json.py
Comment (daviewales) [reply visidata/loaders/json.py]: Done.

File: tests/save-jsonla.vdj
Comment (daviewales) [reply tests/save-jsonla.vdj:2]: I've updated the test to type the columns. It seems to work in batch mode, just not in play mode.

File: tests/load-jsonla.vdj
Comment (daviewales) [reply tests/load-jsonla.vdj]: Done.

**Comment by daviewales** [comment]
Thanks @saulpw, I've pushed some commits with the requested changes.

File: tests/load-jsonla.vdj
Comment (saulpw) [inline tests/load-jsonla.vdj:3]: ah, here I meant also to not open the columns sheet :)  I'll fix this after merging.

**Pull request merged**
